### ISSUE TITLE
GHM 訓練及視覺化分析步驟詳細計畫

### ISSUE THREAD

**Issue by sk413025** [sk413025/audio-angle-classification#19]
# GHM 訓練及視覺化分析步驟詳細計畫

本 issue 提供 Gradient Harmonizing Mechanism (GHM) 訓練實施與視覺化分析的完整步驟指南。

### 1. 準備階段
1. **環境設置**:
   - 確保已安裝所有依賴 (PyTorch, NumPy, Matplotlib 等)
   - 確認 GPU 是否可用並正確設置

2. **數據準備**:
   - 確認音頻數據已按角度類別 (6度) 組織
   - 檢查各頻率 (500Hz, 1000Hz, 3000Hz) 的數據是否齊全
   - 確保數據格式符合 train.py 的要求

### 2. 訓練執行
1. **基礎測試**:
   ```bash
   # 快速測試以確保流程正常 (2個 epoch)
   python train.py --frequency 1000hz --loss-type ghm --ghm-bins 10 --ghm-alpha 0.75 --epochs 2 --seed 42
   ```

2. **完整訓練計劃**:
   ```bash
   # GHM 訓練 (1000Hz, 30個 epoch)
   python train.py --frequency 1000hz --loss-type ghm --ghm-bins 10 --ghm-alpha 0.75 --epochs 30 --seed 42
   
   # 標準損失訓練 (1000Hz, 30個 epoch) - 作為比較基準
   python train.py --frequency 1000hz --loss-type standard --epochs 30 --seed 42
   
   # 其他頻率訓練 (500Hz)
   python train.py --frequency 500hz --loss-type ghm --ghm-bins 10 --ghm-alpha 0.75 --epochs 30 --seed 42
   
   # 其他頻率訓練 (3000Hz)
   python train.py --frequency 3000hz --loss-type ghm --ghm-bins 10 --ghm-alpha 0.75 --epochs 30 --seed 42
   ```

3. **參數探索** (可選):
   ```bash
   # 調整 GHM bins
   python train.py --frequency 1000hz --loss-type ghm --ghm-bins 5 --ghm-alpha 0.75 --epochs 30 --seed 42
   python train.py --frequency 1000hz --loss-type ghm --ghm-bins 20 --ghm-alpha 0.75 --epochs 30 --seed 42
   
   # 調整 alpha 參數
   python train.py --frequency 1000hz --loss-type ghm --ghm-bins 10 --ghm-alpha 0.5 --epochs 30 --seed 42
   python train.py --frequency 1000hz --loss-type ghm --ghm-bins 10 --ghm-alpha 0.9 --epochs 30 --seed 42
   ```

### 3. 訓練監控
1. **記錄訓練輸出**:
   ```bash
   python train.py --frequency 1000hz --loss-type ghm --ghm-bins 10 --ghm-alpha 0.75 --epochs 30 --seed 42 | tee ghm_1000hz_training.log
   ```

2. **從輸出提取關鍵路徑**:
   - 模型檢查點路徑: `<path_to_checkpoint_dir>/model_epoch_<N>.pt`
   - 訓練歷史圖表: `<path_to_plots_dir>/training_history_<timestamp>.png`
   - GHM 統計數據: `<path_to_stats_dir>/epoch*_batch*_<timestamp>_ghm_stats.pkl`
   - 梯度分布圖: `<path_to_plots_dir>/grad_dist_epoch*_batch*_<timestamp>.png`

### 1. 基本分析
1. **訓練歷史分析**:
   - 檢視 `training_history_<timestamp>.png` 了解損失和準確率的趨勢
   - 比較 GHM 和標準損失的收斂速度和最終性能

2. **梯度分布分析**:
   - 檢視 `grad_dist_epoch*_batch*_<timestamp>.png` 了解 GHM 如何平衡梯度分布
   - 比較訓練初期和後期的梯度分布變化

### 2. 詳細 GHM 分析
使用 Python 腳本進行更深入的分析:

```python
import os
import pickle
import matplotlib.pyplot as plt
from utils.visualization import visualize_ghm_stats, compare_ghm_epochs, analyze_ghm_training

# 路徑應替換為訓練輸出中的實際路徑
stats_dir = "<path_to_stats_dir>/plastic_1000hz_ghm_<timestamp>"
plots_dir = "<path_to_plots_dir>/plastic_1000hz_ghm_<timestamp>"
output_dir = "ghm_analysis_results"

os.makedirs(output_dir, exist_ok=True)

# 1. 分析單個 epoch 的 GHM 統計數據
early_epoch = os.path.join(stats_dir, "epoch1_batch0_<timestamp>_ghm_stats.pkl")
mid_epoch = os.path.join(stats_dir, "epoch15_batch0_<timestamp>_ghm_stats.pkl")
late_epoch = os.path.join(stats_dir, "epoch30_batch0_<timestamp>_ghm_stats.pkl")

visualize_ghm_stats(early_epoch, os.path.join(output_dir, "early_epoch_ghm.png"))
visualize_ghm_stats(mid_epoch, os.path.join(output_dir, "mid_epoch_ghm.png"))
visualize_ghm_stats(late_epoch, os.path.join(output_dir, "late_epoch_ghm.png"))

# 2. 比較不同 epoch 的 GHM 統計
compare_ghm_epochs(
    stats_dir,
    epochs=[1, 10, 20, 30],  # 選擇具代表性的 epoch
    output_path=os.path.join(output_dir, "epoch_comparison.png")
)

# 3. 分析整個訓練過程
analyze_ghm_training(
    stats_dir,
    output_dir=os.path.join(output_dir, "full_training_analysis")
)
```

### 3. 性能對比分析
1. **準確率比較**:
   - 比較不同頻率 (500Hz, 1000Hz, 3000Hz) 下的 GHM 性能
   - 比較不同 GHM 參數 (bins, alpha) 對結果的影響
   - 與標準損失函數對比

2. **困難樣本分析**:
   ```python
   # 分析哪些角度類別最難分類
   from sklearn.metrics import confusion_matrix
   import numpy as np
   import torch
   
   # 載入模型和測試數據
   model = torch.load("<path_to_checkpoint_dir>/model_epoch_30.pt")
   # 準備測試數據集並評估模型
   # ...
   
   # 計算混淆矩陣
   conf_matrix = confusion_matrix(y_true, y_pred)
   
   # 繪製混淆矩陣熱圖
   plt.figure(figsize=(10, 8))
   plt.imshow(conf_matrix, cmap='Blues')
   plt.xlabel('Predicted Label')
   plt.ylabel('True Label')
   plt.title('Confusion Matrix')
   plt.colorbar()
   plt.savefig(os.path.join(output_dir, "confusion_matrix.png"))
   ```

### 4. 綜合可視化報告
1. **創建比較表格**:
   - 整理各配置的最佳準確率、損失值、收斂 epoch 等
   - 製作參數影響的條形圖比較

2. **GHM 效果視覺化**:
   - 創建 GHM 權重分布隨時間變化的動畫
   - 展示 GHM 如何重新分配樣本權重

3. **最終報告生成**:
   - 整合所有視覺化結果到一個綜合報告
   - 提出最佳 GHM 配置建議和未來改進方向

## 預期輸出與分析重點

1. **GHM vs 標準損失的主要差異**:
   - 訓練收斂速度比較
   - 最終性能對比
   - 對難分類樣本的處理能力

2. **不同頻率的表現分析**:
   - 哪個頻率對角度分類最有效
   - GHM 在不同頻率數據上的適用性

3. **GHM 參數敏感度**:
   - bins 參數對性能的影響
   - alpha 參數對權重分配的影響
   - 最佳參數組合建議

4. **實踐指南**:
   - 最佳實踐建議
   - 潛在改進方向
   - 在其他音頻分類任務中應用 GHM 的建議

## 時間規劃

1. 基礎測試與數據準備: 1 天
2. 完整訓練執行 (所有配置): 2-3 天
3. 基本視覺化分析: 1 天
4. 詳細 GHM 分析: 1-2 天
5. 綜合報告生成: 1 天

總計: 約 6-8 天完成完整的 GHM 訓練與分析流程

**Comment by sk413025** [comment]
此計畫已成功完成並實施。主要成果包括：

1. GHM 訓練流程已完整實現：
   - 完成了所有頻率 (500Hz, 1000Hz, 3000Hz) 的訓練
   - 實現了參數探索 (bins, alpha) 的比較實驗
   - 建立了完整的訓練監控系統

2. 視覺化分析工具已開發完成：
   - 實現了訓練歷史分析
   - 完成了梯度分布分析功能
   - 建立了 GHM 統計分析工具

3. 代碼組織已優化：
   - 所有分析工具已模組化
   - 建立了清晰的目錄結構
   - 完善了文檔和使用說明

所有相關改進已通過 PR #24 合併到主分支。關閉此 issue。
